soc: atmel: saml2x: Don't do a GCLK SWRST

ringlej opened the pull request:
When using mcuboot, the z_arm_platform_init is executed. When an application is started by mcuboot, it will hang if z_arm_platform_init is run again.

This change will cause the z_arm_platfrom_init code to only be run if CONFIG_BOOTLOADER_MCUBOOT is not set

nandojve commented:
Hi @ringlej ,

The `z_arm_platform_init` must be executed each time a zephyr image boots. This means that if MCUboot is compiled with Zephyr it will execute. The same will happen again with the Zephyr application that runs after. This MUST be the normal behavior because Zephyr application do not necessary depends on MCUboot or any other bootloader and should proper configure their clocks. The MCUboot/Application only works because there is a default clock already configured after reset. So, if the application hangs after MCUboot this only means that the clock sequence is still wrong and should be fixed.

I think the [soc_samd2x](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/atmel/sam0/common/soc_samd2x.c) is already Ok in this regards and you only need to perform the same steps.

ringlej commented:
Hi @nandojve,

Our board has a WLR089U (samr34). MCUboot is doing the `z_arm_platform_init`, but when MCUboot starts an application it, will hang and the application will not start. Specifically, it will have an issue with the call to `gclk_reset()` and the call to `dfll48m_init()`, where if either of those are called when the application starts after MCUboot, it will hang.

The way that I've done the config, if an application is built standalone (without MCUboot), the `CONFIG_SOC_ATMEL_PLATFORM_INIT_ONLY_ONCE` will be set because in this case, `CONFIG_BOOTLOADER_MCUBOOT` will not be defined, so the `z_arm_platform_init()` call will do all of its code. The code is only excluded if an Application is built with `CONFIG_BOOTLOADER_MCUBOOT` set, in which case MCUboot has already taken care of the platform init code and there is no need to do it again (and is problematic at least on a samr34). This issue can be see on the WLR089U Xplained board as well.

Commit:
soc: atmel: saml2x: Don't do a GCLK SWRST

Doing a `GCLK->CTRLA.bit.SWRST = 1` will cause boot chaining to hang.
Setting the CPU clock to run from OSCULP32K during initialization is
all that is needed.

Signed-off-by: Jon Ringle <jringle@gridpoint.com>

ringlej commented:
@attie-argentum Thanks for the pointer.
I compared the samd2x code to the saml2x, and found that an important difference was that the saml2x code is doing a software reset of the clocks which the samd2x is not doing. When I removed the clock software reset, it works correctly. There was already code after the removed clock software reset to set the cpu clock to a safe setting to run from OSCULP32K as the clock source.

attie-argentum commented:
> When I removed the clock software reset, it works correctly

Great!

Have you confirmed correct operation with and without MCUboot?

ringlej commented:
> Have you confirmed correct operation with and without MCUboot?

Yes, I have confirmed that it works both with and without MCUboot

nandojve reviewed (approved):
LGTM since was tested on HW

attie-argentum reviewed (approved):
LGTM, thanks!
